Document import assertions (#19588)
mdn/content: files/en-us/web/javascript/reference/operators/import/index.md

```js
import(moduleName)
import(moduleName, options)
```

The `import()` call is a syntax that closely resembles a function call, but `import` itself is a keyword, not a function. You cannot alias it like `const myImport = import`, which will throw a {{jsxref("SyntaxError")}}.

> [Trailing commas](/en-US/docs/Web/JavaScript/Reference/Trailing_commas) are only allowed if the runtime also supports `options`. Check [browser compatibility](#browser_compatibility).

Reviewer: If I look at the bcd table, all engines listed there support the `options` parameter, don't they?

Author: Yes, this is for older versions, not engines that don't support them at all. It's still relatively new.

Reviewer: Ha! I didn't check bcd PRs… 

I think that _trailing commas_ and `options` are two different (sub-) features. Even if it is likely, there is no guarantee that other engines will implement them simultaneously. (So I think we should update bcd too)

Also, there is no conceptual link between both features, so this would simplify the prose. We could have something as simple as:
```suggestion
As this syntax mimics function calls, [trailing commas](/en-US/docs/Web/JavaScript/Reference/Trailing_commas) {{experimental_inline}} are allowed.
```

(The badge will draw attention to the bcd table below and can be removed when the syntax won't be experimental anymore, that is implemented inside 2 engines at least)

Author: I don't really believe someone would only implement trailing commas without implementing `options`. The current grammar forbids trailing commas (see https://github.com/tc39/proposal-dynamic-import/issues/67), and it's only through adding a second parameter that makes trailing commas actually seem worth it. If you are an engine implementer and you are changing the parser anyway, why not do everything in one? (You can even make it only parsable but throws at runtime.) Otherwise you would end up with a parser that neither adheres to ES2022 nor the import-assertions proposal.

Reviewer: Imagine we are in 5 years. How would know somebody looking at the bcd table know that support for the `options` also means support for trailing commas

Or how will you add the table on the trailing commas page to say when supports for trailing commas on `import` have been added?

Author: I just noticed there's a `javascript.grammar.trailing_commas` section in BCD. In this case I would add a BCD entry, yes. However, I think the content is good as-is—it reduces the cognitive overload of someone trying to understand when's okay to use trailing comma and when's not.

Reviewer 2: Remember you'll need this experimental inline if the feature is present in BCD.

Note I am also assuming this statement is correct as per the spec - which will need to say something like "trailing commas are only allowed if there is more than one option specified to the keyword."

### Parameters

- `moduleName`
  - : The module to import from. The evaluation of the specifier is host-specified, but always follows the same algorithm as static [import declarations](/en-US/docs/Web/JavaScript/Reference/Statements/import).
- `options`

>   - : An object containing import options, which correspond to the appendices of the [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) statement. The follow keys are recognized:

Reviewer: ```suggestion
  - : An object containing import options, which correspond to the appendices of the [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) statement. The following key is recognized:
```

```js-nolint
import(moduleName)
import(moduleName, options)
```

The `import()` call is a syntax that closely resembles a function call, but `import` itself is a keyword, not a function. You cannot alias it like `const myImport = import`, which will throw a {{jsxref("SyntaxError")}}.

[Trailing commas](/en-US/docs/Web/JavaScript/Reference/Trailing_commas) are only allowed if the runtime also supports `options`. Check [browser compatibility](#browser_compatibility).

### Parameters

- `moduleName`
  - : The module to import from. The evaluation of the specifier is host-specified, but always follows the same algorithm as static [import declarations](/en-US/docs/Web/JavaScript/Reference/Statements/import).
- `options`

>   - : An object containing import options, which correspond to the afterthoughts of the [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) statement. The following key is recognized:

Reviewer 2: 1. I don't know that an afterthought is. I went to the linked item and I am still no wiser. Needs to be clear somewhere.
2. Not sure what you can do about it, but I find it confusing that we have two things called `import`.

Reviewer 2: Or to put it another way, if you remove the text ", which correspond to the afterthoughts of the [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) statement" it is easier to work out is going on.

Author: I made this a bit more generic, because there are other proposals that add more stuff after the `import`—like [evaluator attributes](https://github.com/tc39/proposal-import-assertions#follow-up-proposal-evaluator-attributes):

```js
import foo from "mod" with { transformA: "value" };
```

I don't know what to call `assert` and `with` collectively. Afterthoughts? Attributes? Appendices?

Author: > 2\. I find it confusing that we have two things called `import`.

I wasn't happy with the dynamic import's title either (although I created that page). I will change it to `import()`.

Reviewer 2: > I don't know what to call assert and with collectively. Afterthoughts? Attributes? Appendices?

Of those, attributes is best - but I still don't like it. I was thinking "options" or "additional options". If it works, "constraints", "modifiers", "additional specifiers". The problem with afterthoughts is that an afterthought is something almost incidental, and these are significant.

Use dynamic import only when necessary. The static form is preferable for loading initial dependencies, and can benefit more readily from static analysis tools and [tree shaking](/en-US/docs/Glossary/Tree_shaking).

If your file is not run as a module (if it's referenced in an HTML file, the script tag must have `type="module"`), you will not be able to use static import declarations. On the other hand, the asynchronous dynamic import syntax is always available, allowing you to import modules into non-module environments.

> The `options` parameter allows different kinds of import options. For example, [import assertions](/en-US/docs/Web/JavaScript/Reference/Statements/import/assert):

Reviewer 2: So here you've used "options" instead of afterthought right?

Author: “Afterthoughts" is the syntax (`assert`, `with`, etc.); "options" is what they represent. Maybe that sounds a bit arbitrary?

Reviewer 2: It does to me. If you go through and replace afterthoughts with _options_ or _import/export options_ does it read more naturally to you?

Author: I ended up using "attributes" because "options" still sounds weird, given they are written as separate things.

> If your file is not run as a module (if it's referenced in an HTML file, the script tag must have `type="module"`), you will not be able to use static import declarations, but the asynchronous dynamic import syntax will always be available, allowing you to import modules into non-module environments.

Reviewer 2: Would be good to chop this up while we're here.
```suggestion
If your file is not run as a module (if it's referenced in an HTML file, the script tag must have `type="module"`), you will not be able to use static import declarations.
The asynchronous dynamic import syntax will always be available, allowing you to import modules into non-module environments.
```
